src/fsharp/FSharp.Core/array.fs

```diff
@@ -481,36 +481,30 @@ namespace Microsoft.FSharp.Collections
                 | None -> loop(i+1)
                 | res -> res
             loop 0 
-
+        
         [<CompiledName("Choose")>]
         let choose f (array: _[]) = 
             checkNonNull "array" array
-            let res = List<_>() // ResizeArray
+            let res = Array.zeroCreateUnchecked array.Length 
+            let mutable count = 0
             for i = 0 to array.Length - 1 do 
-                match f array.[i] with 
+                match f array.[i] with
                 | None -> ()
-                | Some b -> res.Add(b)
-            res.ToArray()
-
+                | Some b -> res.[count] <- b
+                count <- count + 1
```

Comment:
Am I misreading or this increments count on every pass of the loop? Did you run unit tests for this?